File: src/day07/main.exs
defmodule Day07 do
  def process_line(line, {cd, sizes}) do
    case String.split(line, ~r/\s+/, trim: true) do
      ["$", "cd", ".."] ->
        {tl(cd), sizes}

      ["$", "cd", "/"] ->
        {[], sizes}

      ["$", "cd", name] ->
        {[name | cd], sizes}

      ["$", "ls"] ->
        {cd, sizes}

      ["dir", name] ->
        {cd, Map.put(sizes, [name | cd], 0)}

      [s, _] ->
        {bytes, _} = Integer.parse(s)
        {cd, Day07.update_sizes(cd, bytes, sizes)}
    end
  end

  def update_sizes(path, bytes, sizes) do
    if Enum.empty?(path) do
      Map.update!(sizes, path, &(&1 + bytes))
    else
      update_sizes(tl(path), bytes, Map.update!(sizes, path, &(&1 + bytes)))
    end
  end
end

{_, data} =
  File.read!("puzzle_input/day07")
  |> String.trim_trailing()
  |> String.split(~r/\R+/, trim: True)
  |> Enum.reduce({[], %{[] => 0}}, &Day07.process_line/2)

p1 =
  Map.values(data)
  |> Enum.filter(&(&1 < 100_000))
  |> Enum.sum()

IO.puts("part 1: #{p1}")

to_free = 30_000_000 - 70_000_000 + Map.get(data, [])

p2 =
  Map.values(data)
  |> Enum.filter(&(&1 >= to_free))
  |> Enum.min()

IO.puts("part 2: #{p2}")

File: src/day07/main.rs
#![allow(unused_imports)]
use std::collections::HashMap;

use itertools::Itertools;
use shared::puzzle_input;

const TARGET: usize = 30000000;

pub fn main() {
    let mut sizes = HashMap::from([(vec!["/"], 0)]);
    let mut space = 70000000;
    let mut cd = vec![];
    let input = puzzle_input!();
    for line in input.split('\n') {
        let v = line.split_whitespace().collect_vec();
        match v.as_slice() {
            ["$", "cd", ".."] => {
                cd.pop();
            }
            ["$", "cd", "/"] => cd.clear(),
            ["$", "cd", s] => cd.push(*s),
            ["$", "ls"] => (),
            ["dir", s] => {
                let mut path = cd.clone();
                path.push(s);
                sizes.insert(path, 0);
            }
            [s, _] if s.chars().all(|c| c.is_digit(10)) => {
                let file_size: usize = s.parse().unwrap();
                let mut path = cd.clone();
                while !path.is_empty() {
                    let x = sizes.get_mut(&path).unwrap();
                    *x += file_size;
                    path.pop();
                }

                space -= file_size;
            }
            _ => panic!(),
        }
    }

    let p1: usize = sizes.values().filter(|&&x| x <= 100000).sum();

    let needed = TARGET - space;
    let p2 = sizes.values().filter(|&&x| x >= needed).min().unwrap();

    println!("part 1: {}\npart 2: {}", p1, p2);
}
